docs: follow up to getting started CC and use-cache API ref (#85582)
vercel/next.js: docs/01-app/01-getting-started/06-cache-components.mdx

    - app/api-reference/functions/updateTag
---

> **Good to know:** Cache Components is an opt-in feature. Enable it by setting the `cacheComponents` flag to `true` in your Next config file. See [Enabling Cache Components](#enabling-cache-components) for more details.

Cache Components lets you mix static, cached, and dynamic content in a single route, giving you the speed of static sites with the flexibility of dynamic rendering.

> Server-rendered applications typically force a choice between static pages (fast but stale) and dynamic pages (fresh but slow). Moving this work to the client trades server load for larger bundles and slower initial rendering.

Reviewer: Very well explained!

    - app/api-reference/functions/updateTag
---

> **Good to know:** Cache Components is an opt-in feature. Enable it by setting the `cacheComponents` flag to `true` in your Next config file. See [Enabling Cache Components](#enabling-cache-components) for more details.

> Cache Components lets you build pages that load instantly while still showing fresh, personalized content. You get fine-grained control over what content is rendered and how long it's cached, from build time to request time, giving you both speed and flexibility.

Reviewer: This is good, but it only speaks about the benefits and not what cache components are. Maybe something like: 

In Next.js, you can cache the **return value** of functions and Server Components, giving you fine-grained control over what content is prerendered and reused.

Instead of treating entire pages or apps as static or dynamic, you break down the UI into smaller _units_, or components, [much like in React](https://react.dev/learn/thinking-in-react). Then for those that read data, decide which ones should be cached and for how long, or if they should render at request time and stream.

This page will explain how the Cache Components model works, and how you can start using them.

## Cache Components

> When developing server rendered applications, you have to balance two approaches:

Reviewer: I like how you're introducing the challenges of server-rendering + building apps in the apps here. It's a good way to set the stage for how cache components aim to solve them.

- **Fully static pages** load fast but can't show personalized or real-time data
- **Fully dynamic pages** show fresh data but render everything on each request, leads to slower page loads

However, in practice most pages contain a mix of content types, ranging from static to personalized or real-time data.

Cache Components uses **Partial Prerendering**, React `Suspense` boundaries and Next.js `use cache` directives, to let you compose these different content types into a single page.

- **Partial Prerendering**: At build time, Next.js generates a static HTML shell that includes static content, cached content, and fallback UI for dynamic sections
- **`use cache` directive** : Lets you cache content that doesn't change frequently, so it can be included in the static HTML shell alongside static content, avoiding the need for runtime fetching

> - **Streaming** with `Suspense`: For work that can't complete at build time, defines fallback UI that shows instantly as part of the static HTML shell, while the actual content streams in as it becomes ready

Reviewer: I think we're introducing these concepts way too soon, you could probably cut this section and introduce the concepts as you explain how it works.

  height="632"
/>

## How rendering works with Cache Components

At build time, Next.js renders your route's component tree. As long as components don't access network resources, certain system APIs, or require an incoming request to render, their output is **automatically added to the static shell**. Otherwise, you must choose how to handle them:

> - Defer to request time by wrapping components in React's [`<Suspense>`](https://react.dev/reference/react/Suspense), [showing fallback UI](#defer-rendering-to-request-time) until the content is ready, or

Reviewer: Defer rendering?

```suggestion
- Defer rendering to request time by wrapping components in React's [`<Suspense>`](https://react.dev/reference/react/Suspense), [showing fallback UI](#defer-rendering-to-request-time) until the content is ready, or
```

Next.js requires you to explicitly handle components that can't complete during prerendering. If they aren't wrapped in `<Suspense>` or marked with `use cache`, you'll see an [`Uncached data was accessed outside of <Suspense>`](https://nextjs.org/docs/messages/blocking-route) error during development and build time.

> **Good to know**: Caching can be applied at the component or function level, while fallback UI can be defined around any subtree, which means you can compose static, cached, and dynamic content within a single route.

<Image
  alt="Diagram showing partially rendered page on the client, with loading UI for chunks that are being streamed."
  srcLight="/docs/light/server-rendering-with-streaming.png"
  srcDark="/docs/dark/server-rendering-with-streaming.png"
  width="1600"
  height="785"
/>

> This is how Cache Components implements **Partial Prerendering**. With Cache Components this is the only prerendering strategy. Throughout this document, we refer to it simply as "prerendering".

Reviewer: ```suggestion
This rendering approach is called **Partial Prerendering**, and it's the default behavior with Cache Components. For the rest of this document, we simply refer to it as "prerendering" which can produce a partial or complete output. 
```

> **🎥 Watch:** Why Partial Prerendering and how it works → [YouTube (10 minutes)](https://www.youtube.com/watch?v=MTcPrTIBkpA).

## Automatically prerendered content

Operations like immediately resolved promises, synchronous I/O, module imports, and pure computations can complete during prerendering. Components using only these operations have their rendered output included in the static HTML shell.

Because all operations in the `Page` component below complete during rendering, its rendered output is automatically included in the static shell. When both the layout and page prerender successfully, the entire route is the static shell.

```tsx filename="page.tsx"
import fs from 'node:fs'

export default async function Page() {
  // Immediately resolved promise

>   const data = await new Promise((resolve) => resolve('foo'))

Reviewer: Could you think of a real-world example for this?

      <p>{date.getTime()}</p>
      <p>{uuid}</p>
      <p>{bytes}</p>
    </div>
  )
}
```

Because the `UniqueContent` component defers to request time, to use it within a route, it must be wrapped in `<Suspense>`:

```tsx filename="page.tsx"
export default async function Page() {
  return (
    // <p>Loading..</p> is part of the static shell
    <Suspense fallback={<p>Loading..</p>}>
      <UniqueContent />
    </Suspense>
  )
}
```

Every incoming request would see different random numbers, date, etc.

> **Good to know**: You can cache non-deterministic operations with `use cache`. See the [with non-deterministic operations](#with-non-deterministic-operations) section for examples.

## Using `use cache`

The [`use cache`](/docs/app/api-reference/directives/use-cache) directive caches the return value of async functions and components. You can apply it at the function, component, or file level.

Arguments and any closed-over values from parent scopes automatically become part of the [cache key](/docs/app/api-reference/directives/use-cache#cache-keys), which means different inputs produce separate cache entries. This enables personalized or parameterized cached content.

> When [dynamic content](#dynamic-content) doesn't need to be fetched fresh from the source on every access, caching it lets you include the content in the static shell during prerendering, or reuse the result at runtime across multiple requests.

Reviewer: request instead of access here and below? Trying to reduce variants.

```suggestion
When [dynamic content](#dynamic-content) doesn't need to be fetched fresh from the source on every request, caching it lets you include the content in the static shell during prerendering, or reuse the result at runtime across multiple requests.
```

Cached content can be revalidated in two ways: automatically based on the cache lifetime, or on-demand using tags with [`revalidateTag`](/docs/app/api-reference/functions/revalidateTag) or [`updateTag`](/docs/app/api-reference/functions/updateTag).

> > **Good to know**: See [serialization requirements and constraints](/docs/app/api-reference/directives/use-cache#constraints) in the `use cache` API reference for details on what can be cached and how arguments work.

Reviewer: ```suggestion
> **Good to know**: See [serialization requirements and constraints](/docs/app/api-reference/directives/use-cache#constraints) for details on what can be cached and how arguments work.
```

### During prerendering

While [dynamic content](#dynamic-content) is fetched from external sources, it's often unlikely to change between accesses. Product catalog data updates with inventory changes, blog post content rarely changes after publishing, and analytics reports for past dates remain static.

If this data doesn't depend on [runtime data](#runtime-data), you can use the `use cache` directive to include it in the static HTML shell. Use [`cacheLife`](/docs/app/api-reference/functions/cacheLife) to define how long to use the cached data.

When revalidation occurs, the static shell is updated with fresh content. See [Tagging and revalidating](#tagging-and-revalidating) for details on on-demand revalidation.

```tsx filename="app/page.tsx" highlight={1,4,5}
import { cacheLife } from 'next/cache'

> async function fetchUsers() {

Reviewer: Minor opinion, I think we should lead with caching route segments or components and reserve functions for special cases.

  height="632"
/>

The result is a page that responds instantly with a fast static shell, while dynamic content streams in progressively.

<Image
  alt="Diagram showing partially rendered page on the client, with loading UI for chunks that are being streamed."
  srcLight="/docs/light/server-rendering-with-streaming.png"
  srcDark="/docs/dark/server-rendering-with-streaming.png"
  width="1600"
  height="785"
/>

> **🎥 Watch:** Why PPR and how it works → [YouTube (10 minutes)](https://www.youtube.com/watch?v=MTcPrTIBkpA).

> ## Understanding the Prerendering Flow

Reviewer: Maybe? 

```suggestion
## How rendering works in Next.js
```

Reviewer: I also like how you're essentially explaining Partial Prerendering. So you could probably end this section with: 

`In Next.js, this rendering strategy is called **Partial Prerendering**.`

Maybe not for this page, but there's also an argument for "It's all just PPR". That's the default. You don't have to choose between SSR, SSG, Static, Dynamic, etc. In fact, if you can understand **prerendering** and **run time rendering**, you don't even have to think about the historical rendering strategies.

Author: Should we say, `How rendering works with Cache Components` though

Next.js attempts to prerender everything during the build phase. When it encounters work that cannot complete during prerendering (like `fetch()`, `cookies()`, etc.), that work is deferred to request time:

> 1. **Prerendering starts**: Next.js tries to render the entire page

Reviewer: ```suggestion
1. **Prerendering starts**: Next.js tries to render the entire route
```

2. **Hits dynamic work**: Cannot complete at build time
3. **Requires fallback**: Must be wrapped in `<Suspense>` to define the fallback UI to use in the static shell
4. **Result**: Static shell with clearly defined dynamic sections

Dynamic parts must be explicitly wrapped in `<Suspense>`, or you can use [`use cache`](/docs/app/api-reference/directives/use-cache) to include them in the static shell instead of streaming. If you forget, you'll see an [`Uncached data was accessed outside of <Suspense>`](https://nextjs.org/docs/messages/blocking-route) error during development, and builds will fail.

## Prerendered content

> Operations like immediately resolved promises, synchronous I/O, module imports, and pure computations can complete during the build step. Components using only these operations have their rendered output included in the static HTML shell.

Reviewer: Do you want to explain what the output of the prerender is? 

- Static HTML shell for initial visit
- RSC Payload for route segments (layouts and pages), for initial and subsequent visits. 

If not, the maybe avoid introducing static HTML shell here, could also use "prerendered output" or result to keep it vague and high-level. 

```suggestion
Operations like immediately resolved promises, synchronous I/O, module imports, and pure computations can complete during the build step. Components using only these operations have their rendered output included in the prerendered output.
```

> The static shell is the prerendered HTML that's sent immediately when the page is requested. This means the browser receives fully rendered content right away. During the static shell generation, Next.js also creates a serialized representation of the component tree called the [RSC Payload](/docs/app/getting-started/server-and-client-components#on-the-server), which enables fast client-side navigation between routes.

Reviewer: Oh, I see you mentioned what the output of prerendering here. Maybe swap it around: 

1. Explain how prerendering works and what the output is. 
2. Explain what conditions are **ok** for prerendering

> In the snippet below, the `Page` component requires no special handling. Because all operations complete at build time, the entire page rendered output is the static shell.

Reviewer: Nice example to show everything 👍🏼

```tsx filename="page.tsx"
import fs from 'node:fs'

export default async function Page() {
  // Immediately resolved promise
  const data = await new Promise((resolve) => resolve('foo'))

  // Synchronous I/O
  const content = fs.readFileSync('./config.json', 'utf-8')

  // Module imports
  const constants = await import('./constants.json')

  // Pure computations
  const processed = JSON.parse(content).items.map((item) => item.value * 2)

  return (
    <div>
      <p>{data}</p>
      <h1>{constants.appName}</h1>
      <ul>
        {processed.map((value, i) => (
          <li key={i}>{value}</li>
        ))}
      </ul>
    </div>
  )
}
```

> > **Good to know**: You can verify that a page was fully prerendered by checking the build output summary. Alternatively see what content was added to the static shell of any page by viewing the page source in your browser.

Reviewer: Let's use `route`, since layout + page = route. 

```suggestion
> **Good to know**: You can verify that a route was fully prerendered by checking the build output summary. Alternatively see what content was added to the static shell of any page by viewing the page source in your browser.
```

> > **Good to know**: You can verify that a page was fully prerendered by checking the build output summary. Alternatively see what content was added to the static shell of any page by viewing the page source in your browser.

Reviewer: Maybe for a deep dive you could link to, it would be great to show developers how to look at devtools. The HTML source for the initial visit, and the RSC Payload split by route segments. I wish there as a better way to identify the payload, E.g. for layouts:

`{"children":{"name":"__PAGE__", …}}`

> When your components perform work that can't complete at build time, Cache Components requires you to explicitly choose how to handle it:

Reviewer: Access / read / receive data?

```suggestion
When your components access or read data that can't resolve at build time, Cache Components requires you to explicitly choose how to handle it:
```

Author: Right, I think I got carried away by timeouts - which are mostly just IO simulating tools when rendering. I was actually reworking some of the wording here, let's see if my updates align better w/ this suggestion

- **Stream at request time**: [Define `<Suspense>` fallback UI](#rendering-at-request-time) to be included in the static shell while the actual content streams in

> - **Include in static shell**: [Use the `use cache` directive](#during-prerendering) for content that doesn't change frequently

Reviewer: ```suggestion
- **Render and stream at request time**: [Define `<Suspense>` fallback UI](#rendering-at-request-time) to be included in the static shell while the actual content streams in
- **Prerender and include in static shell**: [Use the `use cache` directive](#during-prerendering) for content that doesn't change frequently
```

```tsx filename="page.tsx"
import { cookies, headers } from 'next/headers'
import { Suspense } from 'react'

async function Example({ searchParams }) {
  // Accessing request data
  const cookieStore = await cookies()
  const headerStore = await headers()
  const search = await searchParams

  return <div>Not in the static shell</div>
}
```

Wrap these components in `<Suspense>`:

```tsx filename="page.tsx"
export default async function Page(props) {
  return (
    // <p>Loading..</p> is part of the static shell
    <Suspense fallback={<p>Loading..</p>}>
      <Example searchParams={props.searchParams} />
    </Suspense>
  )
}
```

Use [`connection()`](/docs/app/api-reference/functions/connection) if you need to defer to request time without accessing any of the runtime APIs above.

> **Good to know**: Runtime data cannot be cached with `use cache` because it requires request context. Components that access runtime APIs must always be wrapped in `<Suspense>`. However, you can extract values from runtime data and pass them as arguments to cached functions. See the [with runtime data](#with-runtime-data) section for an example.

### Non-deterministic operations

> Operations like `Math.random()`, `Date.now()`, or `crypto.randomUUID()` produce different values each time they execute. To ensure these run at request time (generating unique values per request), Next.js requires you to explicitly signal this intent by calling these operations after dynamic or runtime data access.

Reviewer: Good explanation 👍🏼

You can accept unserializable values as arguments as long as you don't introspect them. However, you can return them. This allows patterns like cached components that accept Server or Client Components as children:

```tsx filename="app/cached-wrapper.tsx" switcher
import type { ReactNode } from 'react'
import { setTimeout } from 'node:timers/promises'

async function getSiteTitle() {
  // Simulate a slow database or API call

>   await setTimeout(1000)

Reviewer 2: this is not a valid way to await a timeout

Author: ~Missing the `node:timers/promise` import - updating...~ or nope, it is there at the top

Reviewer 2: holy shit! today i learned. i thought you meant to use a custom delay function but apparently one is built-in now 🤦‍♂️

Author: hehe but - it might trip someone else too, if they don't copy the entire snippet 🤔 - maybe I can alias the import or something like that ~

  return 'My Website'
}

export async function CachedWrapper({ children }: { children: ReactNode }) {
  'use cache'
  const title = await getSiteTitle()

  // Don't introspect children, just pass it through
  return (
    <div className="wrapper">
      <h1>{title}</h1>
      {children}
    </div>
  )
}
```

```jsx filename="app/cached-wrapper.js" switcher
import { setTimeout } from 'node:timers/promises'

async function getSiteTitle() {
  // Simulate a slow database or API call

>   await setTimeout(1000)

Reviewer 2: ditto, this doesn’t work

Reviewer 2: nvm i’m wrong
